server/src/completions.ts

```diff
@@ -0,0 +1,432 @@
+  const matchArray = wordsTillSymbol.match(/\s+/g)
+  if (!matchArray || matchArray.length < 2) {
+    return
+  }
+
+  const suggestions: CompletionItem[] = getSuggestionForFieldAttribute(
+    blockType,
+    currentLine,
+    document,
+    position,
+  )
+
+  return {
+    items: suggestions,
+    isIncomplete: false,
+  }
+}
+
+export function getAllRelationNamesWithoutAst(
+  lines: Array<string>,
+  currentModelStartLine: number,
+): Array<string> {
+  const modelNames: Array<string> = []
+  for (const [key, item] of lines.entries()) {
+    if (currentModelStartLine === key) {
+      // do not suggest the model name we are currently in
+      continue
+    }
+    if (
+      (item.includes('model') || item.includes('enum')) &&
+      item.includes('{')
+    ) {
+      // found a block
+      const blockType = item.replace(/ .*/, '')
+      const blockName = item.slice(blockType.length, item.length - 1).trim()
+
+      modelNames.push(blockName)
+      // block is at least 2 lines long
+    }
+  }
+  return modelNames
+}
+
+export function getSuggestionsForTypes(
+  foundBlock: MyBlock,
+  lines: Array<string>,
+): CompletionList {
+  // create deep copy
+  const suggestions: CompletionItem[] = klona(corePrimitiveTypes)
+
+  if (foundBlock instanceof MyBlock) {
+    // get all model names
+    const modelNames: Array<string> = getAllRelationNamesWithoutAst(
+      lines,
+      foundBlock.start.line,
+    )
+    suggestions.push(
+      ...toCompletionItems(modelNames, CompletionItemKind.TypeParameter),
+    )
+  }
+
+  return {
+    items: suggestions,
+    isIncomplete: false,
+  }
+}
+
+function removeInvalidFieldSuggestions(
+  supportedFields: Array<string>,
+  block: MyBlock,
+  lines: Array<string>,
+  position: Position,
+): Array<string> {
+  let reachedStartLine = false
+  for (const [key, item] of lines.entries()) {
+    if (key === block.start.line + 1) {
+      reachedStartLine = true
+    }
+    if (!reachedStartLine || key === position.line) {
+      continue
+    }
+    if (key === block.end.line) {
+      break
+    }
+    const fieldName = item.replace(/ .*/, '')
+    if (supportedFields.includes(fieldName)) {
+      supportedFields.filter((field) => field !== fieldName)
+    }
+  }
+  return supportedFields
+}
+
+function getSuggestionForDataSourceField(
+  block: MyBlock,
+  lines: Array<string>,
+  position: Position,
+): CompletionItem[] {
+  // create deep copy
+  const suggestions: CompletionItem[] = klona(supportedDataSourceFields)
+
+  const labels: Array<string> = removeInvalidFieldSuggestions(
+    suggestions.map((item) => item.label),
+    block,
+    lines,
+    position,
+  )
+
+  return suggestions.filter((item) => labels.includes(item.label))
+}
+
+function getSuggestionForGeneratorField(
+  block: MyBlock,
+  lines: Array<string>,
+  position: Position,
+): CompletionItem[] {
+  // create deep copy
+  const suggestions: CompletionItem[] = klona(supportedGeneratorFields)
```

Comment:
Is this deep-copy needed?

Reply:
yes it is, since both  `.map()` and `.filter()` operations change the data inside `supportedGeneratorFields`